sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.hpp

```diff
@@ -164,9 +169,147 @@ struct _ur_platform_handle_t : public _ur_platform {
   // Flags to tell whether various Level Zero platform extensions are available.
   bool ZeDriverGlobalOffsetExtensionFound{false};
   bool ZeDriverModuleProgramExtensionFound{false};
+
+  // Cache UR devices for reuse
+  std::vector<std::unique_ptr<_zer_device_handle_t>> PiDevicesCache;
+  pi_shared_mutex PiDevicesCacheMutex;
+  bool DeviceCachePopulated = false;
+
+  // Check the device cache and load it if necessary.
+  zer_result_t populateDeviceCacheIfNeeded();
+
+  // Return the PI device from cache that represents given native device.
+  // If not found, then nullptr is returned.
+  zer_device_handle_t getDeviceFromNativeHandle(ze_device_handle_t);
 };
 
-using ur_platform_handle_t = _ur_platform_handle_t *;
+struct _ur_device_handle_t : _pi_object {
+  _ur_device_handle_t(ze_device_handle_t Device, zer_platform_handle_t Plt,
+                      zer_device_handle_t ParentDevice = nullptr)
+      : ZeDevice{Device}, Platform{Plt}, RootDevice{ParentDevice},
+        ImmCommandListsPreferred{false}, ZeDeviceProperties{},
+        ZeDeviceComputeProperties{} {
+    // NOTE: one must additionally call initialize() to complete
+    // UR device creation.
+  }
+
+  // The helper structure that keeps info about a command queue groups of the
+  // device. It is not changed after it is initialized.
+  struct queue_group_info_t {
+    enum type {

```

Comment:
Avoid using old unscoped C enum.

Reply:
Sure, but I'll leave it for future code refactoring, since this current PR is just moving existing code and it's already huge.